### ISSUE TITLE
Matching on _ char gives incorrect unreachable pattern 

### ISSUE THREAD

**Issue by gpit2286** [rust-lang/rust-clippy#12207]
### Summary

Gives warning that anything after '_' is unreachable, but works as intended. 

### Reproducer

I tried this code:

```rust
pub fn is_id_start(c: char) -> bool {
    match c {
        'A' ..= 'z' | '_' | '-' => true, 
        _ => false 
    }
}

fn main(){
println!("a: {:?}", is_id_start('a'));
println!("1: {:?}", is_id_start('1'));
println!("_: {:?}", is_id_start('_'));
}
```

Playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d7a7b33f303c142a6f54a83166ae1447

I expected to see this happen:
No warning 

Instead, this happened:
warning: unreachable pattern
   --> src/main.rs:147:23
    |
147 |         'A' ..= 'z' | '_' | '-' => true, 
    |                       ^^^
    |
    = note: `#[warn(unreachable_patterns)]` on by default

### Version

```text
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: aarch64-apple-darwin
release: 1.75.0
LLVM version: 17.0.6
```

**Comment by Jarcho** [comment]
The warning is correct. `_` (0x5f) is between `A` (0x41) and `z` (0x7a).
